src/biokbase/narrative/jobs/job.py

```diff
@@ -240,6 +238,76 @@ def state(self):
         except Exception as e:
             raise Exception(f"Unable to fetch info for job {self.job_id} - {e}")
 
+    def revised_state(self, state=None) -> dict:
+        """
+        :param state: can be queried individually from ee2/cache with self.state(),
+            but sometimes want it to be queried in bulk from ee2 upstream
+        :return: dict, with structure
+
+        {
+            owner: string (username, who started the job),
+            spec: app spec (optional)
+            widget_info: (if not finished, None, else...) job.get_viewer_params result
+            state: {
+                job_id: string,
+                status: string,
+                created: epoch ms,
+                updated: epoch ms,
+                queued: optional - epoch ms,
+                finished: optional - epoc ms,
+                terminated_code: optional - int,
+                tag: string (release, beta, dev),
+                parent_job_id: optional - string or null,
+                run_id: string,
+                cell_id: string,
+                errormsg: optional - string,
+                error (optional): {
+                    code: int,
+                    name: string,
+                    message: string (should be for the user to read),
+                    error: string, (likely a stacktrace)
+                },
+                error_code: optional - int
+            }
+        }
+        """
+        if not state:
+            state = self.state()
+        else:
+            state = self._augment_ee2_state(state)
+        state["child_jobs"] = []
+        batch_size = state.get("job_input", {}).get("narrative_cell_info", {}).get("batch_size")
+        if batch_size:
+            state["batch_size"] = batch_size
+
+        widget_info = None
+        if state.get("finished"):
+            try:
+                widget_info = self.get_viewer_params(state)
+            except Exception as e:
+                # Can't get viewer params
+                new_e = transform_job_exception(e)
+                state.update(
+                    {
+                        "status": "error",
+                        "errormsg": "Unable to build output viewer parameters!",
+                        "error": {
+                            "code": getattr(new_e, "code", -1),
+                            "source": getattr(new_e, "source", "JobManager"),
+                            "name": "App Error",
+                            "message": "Unable to build output viewer parameters",
+                            "error": "Unable to generate App output viewer!\nThe App appears to have completed successfully,\nbut we cannot construct its output viewer.\nPlease contact the developer of this App for assistance.",
```

Comment:
I think we're trying to generally update to "concat the KBase team at https://kbase.us/support for assistance." or something like that.

Reply:
I like the idea of concat-ing the KBase team -- how far do you think we would reach? :wink:

Reply:
Everyone! I think that was an 80's horror film...